File: wpilibj/src/main/java/edu/wpi/first/wpilibj2/command/SendableCommandBase.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2018-2019 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package edu.wpi.first.wpilibj2.command;

import java.util.HashSet;
import java.util.Set;

import edu.wpi.first.wpilibj.Sendable;
import edu.wpi.first.wpilibj.smartdashboard.SendableBuilder;

/**
 * A {@link Sendable} base class for {@link Command}s.
 */
@SuppressWarnings("PMD.AbstractClassWithoutAbstractMethod")
public abstract class SendableCommandBase implements Sendable, Command {

  protected String m_name = this.getClass().getSimpleName();
  protected String m_subsystem = "Ungrouped";
  protected Set<Subsystem> m_requirements = new HashSet<>();

  /**
   * Adds the specified requirements to the command.
   *
   * @param requirements the requirements to add
   */
  public final void addRequirements(Subsystem... requirements) {
    m_requirements.addAll(Set.of(requirements));
  }

  @Override
  public Set<Subsystem> getRequirements() {
    return m_requirements;
  }

  @Override
  public String getName() {
    return this.getClass().getSimpleName();
  }

  @Override
  public void setName(String name) {
    m_name = name;
  }

  @Override
  public String getSubsystem() {
    return m_subsystem;
  }

  @Override
  public void setSubsystem(String subsystem) {
    m_subsystem = subsystem;
  }

  /**
   * Initializes this sendable.  Useful for allowing implementations to easily extend SendableBase.
   *
   * @param builder the builder used to construct this sendable
   */
  @Override
  public void initSendable(SendableBuilder builder) {
    builder.setSmartDashboardType("Command");
    builder.addStringProperty(".name", this::getName, null);
    builder.addBooleanProperty("running", this::isScheduled, value -> {
      if (value) {
        if (!isScheduled()) {
          schedule();
        }
      } else {
        if (isScheduled()) {
          cancel();
        }
      }
    });
    builder.addBooleanProperty(".isParented",
        () -> CommandGroupBase.getGroupedCommands().contains(this), null);
  }
}

File: wpilibc/src/main/native/cpp/frc2/command/SendableCommandBase.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2019 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include "frc2/command/SendableCommandBase.h"

#include <frc/smartdashboard/SendableBuilder.h>
#include <frc2/command/CommandScheduler.h>
#include <frc2/command/SetUtilities.h>

using namespace frc2;

SendableCommandBase::SendableCommandBase() {
  m_name = Command::GetName();
  m_subsystem = "Unknown";
}

SendableCommandBase::SendableCommandBase(const SendableCommandBase& other)
    : Sendable{}, Command{} {
  m_name = other.m_name;
  m_subsystem = other.m_subsystem;
  m_requirements = other.m_requirements;
}

void SendableCommandBase::AddRequirements(
    std::initializer_list<Subsystem*> requirements) {
  m_requirements.insert(requirements.begin(), requirements.end());
}

void SendableCommandBase::AddRequirements(
    wpi::SmallSet<Subsystem*, 4> requirements) {
  m_requirements.insert(requirements.begin(), requirements.end());
}

wpi::SmallSet<Subsystem*, 4> SendableCommandBase::GetRequirements() const {
  return m_requirements;
}

void SendableCommandBase::SetName(const wpi::Twine& name) {
  m_name = name.str();
}

std::string SendableCommandBase::GetName() const { return m_name; }

std::string SendableCommandBase::GetSubsystem() const { return m_subsystem; }

void SendableCommandBase::SetSubsystem(const wpi::Twine& subsystem) {
  m_subsystem = subsystem.str();
}

void SendableCommandBase::InitSendable(frc::SendableBuilder& builder) {
  builder.SetSmartDashboardType("Command");
  builder.AddStringProperty(".name", [this] { return GetName(); }, nullptr);
  builder.AddBooleanProperty("running", [this] { return IsScheduled(); },
                             [this](bool value) {
                               bool isScheduled = IsScheduled();
                               if (value && !isScheduled) {
                                 Schedule();
                               } else if (!value && isScheduled) {
                                 Cancel();
                               }
                             });
}
